pythonwhat/State.py

```diff
@@ -159,150 +106,11 @@ def get_manual_sigs(self):
-
-        if (self.solution_for_calls is None):
-            fp = ForParser()
-            fp.visit(self.solution_tree)
-            self.solution_for_calls = fp.fors
-
-    def extract_function_defs(self):
-        if (self.student_function_defs is None):
-            fp = FunctionDefParser()
-            fp.visit(self.student_tree)
-            self.student_function_defs = fp.defs
-
-        if (self.solution_function_defs is None):
-            fp = FunctionDefParser()
-            fp.visit(self.solution_tree)
-            self.solution_function_defs = fp.defs
-
-    def extract_lambda_functions(self):
-        if (self.student_lambda_functions is None):
-            lfp = LambdaFunctionParser()
-            lfp.visit(self.student_tree)
-            self.student_lambda_functions = lfp.funs
-
-        if (self.solution_lambda_functions is None):
-            lfp = LambdaFunctionParser()
-            lfp.visit(self.solution_tree)
-            self.solution_lambda_functions = lfp.funs
-
-    def extract_list_comps(self):
-        if self.student_list_comps is None:
-            lcp = ListCompParser()
-            lcp.visit(self.student_tree)
-            self.student_list_comps = lcp.comps
-
-        if self.solution_list_comps is None:
-            lcp = ListCompParser()
-            lcp.visit(self.solution_tree)
-            self.solution_list_comps = lcp.comps
-
-    def extract_dict_comps(self):
-        if self.student_dict_comps is None:
-            dcp = DictCompParser()
-            dcp.visit(self.student_tree)
-            self.student_dict_comps = dcp.comps
-
-        if self.solution_dict_comps is None:
-            dcp = DictCompParser()
-            dcp.visit(self.solution_tree)
-            self.solution_dict_comps = dcp.comps
-
-    def extract_generator_exps(self):
-        if self.student_generator_exps is None:
-            gep = GeneratorExpParser()
-            gep.visit(self.student_tree)
-            self.student_generator_exps = gep.comps
-
-        if self.solution_dict_comps is None:
-            gep = GeneratorExpParser()
-            gep.visit(self.solution_tree)
-            self.solution_generator_exps = gep.comps
-
-    def extract_withs(self):
-        if (self.student_withs is None):
-            wp = WithParser()
-            wp.visit(self.student_tree)
-            self.student_withs = wp.withs
-
-        if (self.solution_withs is None):
-            wp = WithParser()
-            wp.visit(self.solution_tree)
-            self.solution_withs = wp.withs
-
-    def extract_try_excepts(self):
-        if (self.student_try_excepts is None):
-            tep = TryExceptParser()
-            tep.visit(self.student_tree)
-            self.student_try_excepts = tep.try_excepts
-
-        if (self.solution_try_excepts is None):
-            tep = TryExceptParser()
-            tep.visit(self.solution_tree)
-            self.solution_try_excepts = tep.try_excepts
+    @property
```

Comment:
From paring: this was because of the caching issue mentioned below. Getters re-run the parsers. I'll add a cache, so we can use a getter for the FunctionParser properties that still emulates the  previous behavior.